src/fs/filter.rs

```diff
@@ -66,11 +78,22 @@ impl FileFilter {
     /// Remove every file in the given vector that does *not* pass the
     /// filter predicate for files found inside a directory.
     pub fn filter_child_files(&self, files: &mut Vec<File<'_>>) {
+        use FileFilterFlags::{OnlyDirs, OnlyFiles};
+        
         files.retain(|f| ! self.ignore_patterns.is_ignored(&f.name));
 
-        if self.only_dirs {
-            files.retain(File::is_directory);
+        match (self.flags.contains(&OnlyDirs), self.flags.contains(&OnlyFiles)) {
+            (true, false) => {
+                // On pass -'-only-dirs' flag only
+                files.retain(File::is_directory);
+            }
+            (false, true) => {
+                // On pass -'-only-files' flag only
+                files.retain(File::is_file);
+            }
+            _ => {}
```

Comment:
Doing it like this means that `-f --only-dirs` falls through and doesn't just show no files. It's probably the best solution, because `-f --only-dirs` is completely useless if it didn't show anything.